fxrunner/src/lib/request.rs

```diff
@@ -0,0 +1,176 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::borrow::Cow;
+use std::io;
+use std::path::{Path, PathBuf};
+
+use async_trait::async_trait;
+use rand::prelude::*;
+use thiserror::Error;
+use tokio::fs::create_dir;
+
+use crate::fs::PathExt;
+
+#[derive(Clone)]
+pub struct RequestInfo<'a> {
+    pub id: Cow<'a, str>,
+    pub path: PathBuf,
+}
+/// A trait for creating and validating requests.
+#[async_trait]
+pub trait RequestManager {
+    /// Create a new request.
+    async fn new_request(&self) -> Result<RequestInfo<'static>, NewRequestError>;
+
+    /// Attempt to resume a request with the given ID.
+    ///
+    /// If the request ID is valid and a prior request exists, the path to its
+    /// directory will be returned.
+    async fn resume_request<'a>(
+        &self,
+        request_id: &'a str,
+    ) -> Result<RequestInfo<'a>, ResumeRequestError>;
+
+    /// Ensure the profile directory for the given request exists.
+    async fn ensure_profile_dir<'a>(
+        &self,
+        request_info: &RequestInfo<'a>,
+    ) -> Result<PathBuf, io::Error>;
+}
+
+pub struct FsRequestManager {
+    path: PathBuf,
+}
+
+impl FsRequestManager {
+    pub fn new(path: &Path) -> Self {
+        FsRequestManager { path: path.into() }
+    }
+}
+
+#[async_trait]
+impl RequestManager for FsRequestManager {
+    async fn new_request(&self) -> Result<RequestInfo<'static>, NewRequestError> {
+        const TRIES: u64 = 30;
+        const DICT: &[char] = &[
+            'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q',
+            'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h',
+            'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y',
+            'z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+        ];
+
+        let mut request_id = String::with_capacity(8);
+        let mut path = self.path.clone();
+
+        for _ in 0..TRIES {
+            for _ in 0..8 {
+                let mut rng = thread_rng();
+                request_id.push(*DICT.choose(&mut rng).unwrap());
+            }
+
+            path.push(&request_id);
+
+            if let Err(e) = create_dir(&path).await {
+                if e.kind() == io::ErrorKind::AlreadyExists {
+                    request_id.clear();
+                    path.pop();
+                    continue;
+                } else {
+                    return Err(e.into());
+                }
+            }
+
+            return Ok(RequestInfo {
+                path,
+                id: Cow::Owned(request_id),
+            });
+        }
+
+        Err(NewRequestError::TooManyAttempts(TRIES))
+    }
+
+    async fn resume_request<'a>(
+        &self,
+        request_id: &'a str,
+    ) -> Result<RequestInfo<'a>, ResumeRequestError> {
+        if request_id.chars().any(|c| !c.is_ascii_alphanumeric()) {
```

Comment:
(Of course you'd need to remove this bit if you switched to base64 and just check if it failed to decode)